charmcraft/commands/store/__init__.py

```diff
@@ -246,41 +246,67 @@ class WhoamiCommand(BaseCommand):
     """
     )
 
+    def fill_parser(self, parser):
+        """Add own parameters to the general parser."""
+        self.include_format_option(parser)
+
     def run(self, parsed_args):
         """Run the command."""
         store = Store(self.config.charmhub)
         try:
             macaroon_info = store.whoami()
         except CredentialsUnavailable:
-            emit.message("You are not logged in to Charmhub.")
+            if parsed_args.format:
+                info = {"logged": False}
+                emit.message(self.format_content(parsed_args.format, info))
+            else:
+                emit.message("You are not logged in to Charmhub.")
             return
 
-        emit.message(f"name: {macaroon_info.account.name}")
-        emit.message(f"username: {macaroon_info.account.username}")
-        emit.message(f"id: {macaroon_info.account.id}")
+        human_msgs = []
+        prog_info = {"logged": True}
+
+        human_msgs.append(f"name: {macaroon_info.account.name}")
+        prog_info["name"] = macaroon_info.account.name
+        human_msgs.append(f"username: {macaroon_info.account.username}")
+        prog_info["username"] = macaroon_info.account.username
+        human_msgs.append(f"id: {macaroon_info.account.id}")
+        prog_info["id"] = macaroon_info.account.id
 
         if macaroon_info.permissions:
-            emit.message("permissions:")
+            human_msgs.append("permissions:")
             for item in macaroon_info.permissions:
-                emit.message(f"- {item}")
+                human_msgs.append(f"- {item}")
+            prog_info["permissions"] = macaroon_info.permissions
 
         if macaroon_info.packages:
             grouped = {}
             for package in macaroon_info.packages:
                 grouped.setdefault(package.type, []).append(package)
             for package_type, title in [("charm", "charms"), ("bundle", "bundles")]:
                 if package_type in grouped:
-                    emit.message(f"{title}:")
+                    human_msgs.append(f"{title}:")
+                    pkg_info = []
                     for item in grouped[package_type]:
                         if item.name is not None:
-                            emit.message(f"- name: {item.name}")
+                            human_msgs.append(f"- name: {item.name}")
+                            pkg_info.append({"name": item.name})
                         elif item.id is not None:
-                            emit.message(f"- id: {item.id}")
+                            human_msgs.append(f"- id: {item.id}")
+                            pkg_info.append({"id": item.id})
```

Comment:
The spec is sub-defined here. For each package I put a dict instead of just the package name, to indicate which kind of value it holds (the name or the id).